clients/docs/includes/configs/cloud/java-sr.config

```diff
@@ -1,9 +1,13 @@
-# Kafka
+# Kafka client (producer/consumer/admin)
 bootstrap.servers={{ BROKER_ENDPOINT }}
 security.protocol=SASL_SSL
 sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required username="{{ CLUSTER_API_KEY }}" password="{{ CLUSTER_API_SECRET }}";
-ssl.endpoint.identification.algorithm=https
 sasl.mechanism=PLAIN
+client.dns.lookup=use_all_dns_ips
```

Comment:
`client.dns.lookup=use_all_dns_ips` is an optimization, it is not required, and so for reasons already mentioned in this PR, maybe this shouldn't be here either and rather moved to `# Recommended settings for optimal Kafka producer performance`.

Reply:
This is not an optimization, it is required for correctness.

Reply:
Going back to https://github.com/confluentinc/examples/pull/820#discussion_r498894181:

> Original intention was that these were just minimum set of required parameters to _connect_ to a Kafka cluster. 

The only required configurations are the following 4 configuration parameters:

```
bootstrap.servers={{ BROKER_ENDPOINT }}
security.protocol=SASL_SSL
sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required username="{{ CLUSTER_API_KEY }}" password="{{ CLUSTER_API_SECRET }}";
sasl.mechanism=PLAIN
```

Since clients can still work without setting `client.dns.lookup=use_all_dns_ips`, it is not required in these configuration files (let's not bikeshed about the word `optimization` versus `correctness`) nor are the other parameters.

My leaning is still to keep these files to these 4 parameters, and link to Confluent documentation which can describe the other recommendations.

Reply:
@ybyzek I did not intend to bikeshed, I think it's an important distinction. I have been in several incidents due to `use_all_dns_ips` not being set. @Evelyn-Bayes asked a related question a few days ago:

https://confluent.slack.com/archives/CF6E4C48N/p1601611374088300

I am a big fan of keeping configs as simple as we can, but I do think that having unsafe configs by default hurts our users in a big way and it's extremely costly for us. The sad truth is that many of our Cloud users don't read the documentation until it's too late. Often only after an incident that involves GTS, Engineering and causes a serious outage on their end.

I think a case can be made that `linger.ms=5` should be be removed from the PR (even though it costs us real $ and results in a worse experience to our users, particularly in MT clusters) and I can go along with that, but the rest are really essential unless you're an expert user and you know what you're doing.

Reply:
💯 agree that these additional configuration parameters help users have a better experience.  (I did not mean to convey that these additional configuration parameters are not important.)

We should strive to help users in three ways, in order of highest-to-lowest impact:

- Make them the default, like `use_all_dns_ips` which became the default with https://issues.apache.org/jira/browse/KAFKA-9313, which is an Apache Kafka jira that I filed last year
- In-product help (Confluent Cloud UI in this case)
- Out-of-product documentation that lives somewhere else (like docs.confluent.io)

The crux of this PR is how to address #2.  Your point makes sense, so let's find a way to present the information in the Confluent Cloud UI, it will have the broadest reach and greatest potential to get used so that users configure their clients for better behavior. 👍 

In doing so, we have to address one concern: users copy/paste things exactly as shown to them, and don't drill down to understand what the configuration parameters mean.  If the UI shows them X, they will blindly copy and paste X to EVERYTHING: producers, consumers, CLI, etc,...and then they will copy/paste the whole thing again when they want to reuse the properties file to connect to some other cluster...  This causes potential friction points: 

(a) differentiating what is mandatory to connect to Confluent Cloud from what gives better experience
(b) which configs are for all clients versus just producer versus just consumer, etc
(c) portability of properties file -- users may take the properties intended for CCloud and then blindly reuse the same configuration file for testing on a local cluster.

So in whatever way we present these configurations in the Confluent Cloud UI, it needs to be very clear and crisp, explaining what is for what.  I might have argued for different blocks of code with separate copy/paste sections, but that involves more investment on the Confluent Cloud UI side, so the lowest cost is putting everything into the same block as done here.  Let's keep iterating and polishing the ideas we had with the comments in this GH issue.  

@ijuma  **Would you be able to make another pass at the changes in this PR and add embed more elaborate comments (comments into the files themselves, not GH comments) that explain the parameters in more detail that address (a), (b), and (c) above?**

Additionally:

1. **AK Defaults**: it sounds like there is some further possibility for improvement with setting AK defaults for `reconnect.backoff.max.ms ` -- is there an Apache Kafka jira for this yet?

2. @Evelyn-Bayes 's message also mentioned `networkaddress.cache.ttl` and `networkaddress.cache.negative.ttl`, that doesn't fit this model...where does that go into the Confluent Cloud UI?

3. **Documentation enhancement** :  https://github.com/confluentinc/docs/pull/4730/files did not update with all the parameters (including) `reconnect.backoff.max.ms`, and the docs at https://docs.confluent.io/current/cloud/using/config-client.html#recommendations may still need to be updated.  Are you going to fix that?

4. Are you going to file a DOCS jira for all the other components to remove `ssl.endpoint.identification.algorithm=https` and add these recommended configurations?

Reply:
Totally agreed with the way you described the 3 ways we should help users. I am hoping that we can make `acks=all` and `enable.idempotence=true` the default in Apache Kafka 3.0 for better out of the box behavior in that respect too (@d8tltanc is looking into that in Q4).

Yes, I can add descriptions with the configs. Answers to your numbered questions:

1. There isn't a JIRA, but there is a KIP for introducing exponential backoff for retries. Maybe we can consider making the change as part of that. I pinged the relevant people.
2. Not sure. Also, the defaults are actually reasonable unless a security manager is used. https://github.com/openjdk/jdk/blob/6bab0f539fba8fb441697846347597b4a0ade428/src/java.base/share/classes/sun/net/InetAddressCachePolicy.java#L48. So, maybe it's OK for it to be in docs only.
3. Yeah, I submitted https://github.com/confluentinc/docs/pull/5845 last week and I was planning to add Jun's suggestion there.
4. After we reach consensus and we merge this PR, sure.
5. Great! Will review.
6. You mean all ccloud examples or all examples?

Reply:
@ybyzek I removed linger.ms and tweaked the comments a bit. I think it's best to keep it simple instead of trying to explain in a lot of detail. Please let me know what you think.